### PR TITLE
add multi ifo support to pycbc live

### PR DESCRIPTION
**Pull request by ahnitz**
This adds multi ifo support to pycbc live. 
    * Supports arbitrary number of detectors for outputting triggers
    * All ifos contribute to significance of detectors
    * No SNR threshold for all but the two ifos

The basic features are all there, so please review. 

[WIP]
 * formatting / pep8 / some documentation.

Notes:

This removes the "followup" detector options. No detectors can just be followup at the moment, they will always contribute to significance and will have single triggers generated. We may want to add options to ignore some ifos for some purposes, but we can add that in a future patch.

### CODE REVIEW

File: pycbc/events/coinc.py [hunk 1]
```diff
@@ -594,11 +592,22 @@ def __init__(self, num_templates, analysis_block, background_statistic,
         from . import stat
         self.num_templates = num_templates
         self.analysis_block = analysis_block
+
+        # Only pass a valid stat file for this ifo pair
+        for fname in stat_files:
+            f = h5py.File(fname, 'r')
```
Comment (tdent) [inline pycbc/events/coinc.py]: doesn't h5py need to be imported before this?

File: pycbc/events/coinc.py [hunk 2]
```diff
@@ -612,11 +621,60 @@ def __init__(self, num_templates, analysis_block, background_statistic,
...
-        #    def sig_handler(signum, frame):
-        #        pass
-        #    signal.signal(signal.SIGINT, sig_handler)
+    # Adding this here as it requires the results format dict as defined by this
+    # class. 
+    @classmethod
+    def pick_best_coinc(cls, coinc_results):
+        """Choose the best coinc by ifar first, then statistic if needed.
+
+        This function picks which of the available coincs to use. It chooses
```
Comment (tdent) [inline pycbc/events/coinc.py]: 'which.. to use' -> to use for what?  Also might be good to say if this function is used only in multi-ifo time or also in 2-ifo time and say explicitly that each 'coinc' is a 2-ifo coinc.
suggest 'chooses first by ifar ..' -> 'first selects the highest-ifar coinc, then uses highest statistic as a tie-breaker'
trails -> trials
'were possible at this time' -> 'are possible given the active ifos'

File: pycbc/events/coinc.py [hunk 3]
```diff
@@ -612,11 +621,60 @@ def __init__(self, num_templates, analysis_block, background_statistic,
...
+        coinc_results: list of coinc result dicts
+            Dictionary by detector pair of coinc result dicts.
+
+        Returns
+        -------
+        best: coinc results dict
+            Dict containing the results as would be returned by this class.
+            If there is a coinc, this will contain the 'best' one.
+        """
+        # Choose best by ifar, if equal choose loudest statistic value
```
Comment (tdent) [inline pycbc/events/coinc.py]: move this comment to the piece of code that actually does this

File: pycbc/events/coinc.py [hunk 4]
```diff
@@ -612,11 +621,60 @@ def __init__(self, num_templates, analysis_block, background_statistic,
...
+        mstat = 0
+        mifar = 0
+        mresult = None
+
+        # record the trials factor from the possible coincs we could
+        # maximize over
+        trials = 0
+        for result in coinc_results:
+            # Check that a coinc was possible
+            if 'coinc_possible' in result:
```
Comment (tdent) [inline pycbc/events/coinc.py]: where is this 'coinc_possible' string added to the dict?

File: pycbc/events/coinc.py [hunk 5]
```diff
@@ -612,11 +621,60 @@ def __init__(self, num_templates, analysis_block, background_statistic,
...
+                trials += 1
+
+                # Check that a coinc exists
+                if 'foreground/ifar' in result:
+                    ifar = result['foreground/ifar']
```
Comment (tdent) [inline pycbc/events/coinc.py]: is this always exactly one IFAR value?

File: bin/pycbc_live [hunk 6]
```diff
@@ -83,40 +98,72 @@ class LiveEventManager(object):
...
+        triggers['foreground/{}/end_time'.format(ifo)] = float(ptime)
+        triggers['foreground/{}/snr'.format(ifo)] = abs(snr_series.at_time(ptime))
+        triggers['foreground/{}/stat'.format(ifo)] = triggers['foreground/{}/snr'.format(ifo)]
+        triggers['foreground/{}/coa_phase'.format(ifo)] = numpy.angle(snr_series.at_time(ptime))
+        triggers['foreground/{}/chisq'.format(ifo)] = 0
+                     
     def check_coincs(self, ifos, coinc_results, psds, f_low,
-                     data_readers, bank, followup_ifos=None):
+                     data_readers, bank):
         """ Save zerolag triggers to a coinc xml file """
```
Comment (tdent) [inline bin/pycbc_live]: is this all the function does? looks like it also manages the followup, then say so

File: pycbc/events/coinc.py (hunk 5 above)
Comment (ahnitz) [reply pycbc/events/coinc.py]: Yes, by construction. There is only one coincidence per analysis stride (~ 8 seconds).

File: bin/pycbc_live [hunk 7]
```diff
@@ -129,16 +176,15 @@ class LiveEventManager(object):
...
+    def check_singles(self, results, data_reader, psds, f_low):
         active = [k for k in results if results[k] != None]
         if len(active) == 1:
             ifo = active[0]
             single = sngl_estimator[ifo].check(results[ifo], data_reader[ifo])
 
             if single is not None:
                 fud = self.compute_followup_data([ifo], single, data_reader,
-                                                 bank, followup_ifos)
+                                                 bank, [])
```
Comment (titodalcanton) [inline bin/pycbc_live]: Do I understand right that this disables multidetector skymaps for single events? Did the patch break functions in a way that requires this?

File: bin/pycbc_live [hunk 8]
```diff
@@ -356,11 +402,7 @@ if args.size_override:
 
 # ifos used for primary analysis (only two at the moment)
 ifos = set(args.channel_name.keys())
-followup_ifos = set(args.followup_detectors)
-ifos -= followup_ifos
-ifos, followup_ifos = list(ifos), list(followup_ifos)
-# the root must also advance followup ifos, other threads don't need to
-advancing_ifos = (ifos + followup_ifos) if evnt.rank == 0 else ifos
+logging.info('Running with: %s ifos', len(ifos))
```
Comment (titodalcanton) [inline bin/pycbc_live]: Let's be explicit:
```suggestion
logging.info('Running with %s', ', '.join(sorted(ifos)))
```

File: bin/pycbc_live [hunk 9]
```diff
@@ -523,24 +563,31 @@ with ctx:
...
             # Look for coincident triggers and do background estimation
-            coinc_results = {}
+            coinc_results = []
             if args.enable_background_estimation:
-                coinc_results = estimator.add_singles(results, data_reader)
-                evnt.check_coincs(results.keys(), coinc_results,
-                                  psds, args.low_frequency_cutoff, data_reader, bank,
-                                  followup_ifos=usable_followup_ifos)
+                for c in estimators:
+                    coinc_results.append(c.add_singles(results, data_reader))
```
Comment (titodalcanton) [inline bin/pycbc_live]: Will the add_singles method just ignore the ifos that do not concern it?

File: bin/pycbc_live [hunk 10]
```diff
@@ -523,24 +563,31 @@ with ctx:
...
 
             # Check for singles if we don't have coinc time
             if args.enable_single_detector_background:
                 evnt.check_singles(results, data_reader, psds,
-                                   args.low_frequency_cutoff,
-                                   followup_ifos=usable_followup_ifos)
+                                   args.low_frequency_cutoff)
 
+            # map the results file to an hdf file
             prefix = '%s-%s-%s-%s' % (''.join(ifos), args.file_prefix, data_end() - args.analysis_chunk, valid_pad)
```
Comment (titodalcanton) [inline bin/pycbc_live]:
```suggestion
            prefix = '{}-{}-{}-{}'.format(''.join(sorted(ifos)), args.file_prefix, data_end() - args.analysis_chunk, valid_pad)
```

File: pycbc/events/coinc.py [hunk 11]
```diff
@@ -612,11 +621,60 @@ def __init__(self, num_templates, analysis_block, background_statistic,
...
+
+        Parameters
+        ----------
+        coinc_results: list of coinc result dicts
+            Dictionary by detector pair of coinc result dicts.
+
+        Returns
+        -------
+        best: coinc results dict
+            Dict containing the results as would be returned by this class.
```
[the remark this answers is not in the document]
Comment (titodalcanton) [reply pycbc/events/coinc.py]: Typo: trails → trials

File: pycbc/events/coinc.py (hunk 1 above)
Comment (ahnitz) [reply pycbc/events/coinc.py]: Yes, done.

File: bin/pycbc_live [hunk 12]
```diff
@@ -83,40 +98,72 @@ class LiveEventManager(object):
...
+            if 'foreground/{}/'.format(coinc_ifo) in key:
+                _, _, name = key.split('/')
+                triggers['foreground/{}/{}'.format(ifo, name)] = triggers[key]
+
+        # Set the critical stats manually
+        triggers['foreground/{}/end_time'.format(ifo)] = float(ptime)
+        triggers['foreground/{}/snr'.format(ifo)] = abs(snr_series.at_time(ptime))
+        triggers['foreground/{}/stat'.format(ifo)] = triggers['foreground/{}/snr'.format(ifo)]
+        triggers['foreground/{}/coa_phase'.format(ifo)] = numpy.angle(snr_series.at_time(ptime))
+        triggers['foreground/{}/chisq'.format(ifo)] = 0
```
[the remark this answers is not in the document]
Comment (ahnitz) [reply bin/pycbc_live]: I think setting both to zero makes sense here. This is only for the coinc xml file, which the detchar summary pages don't use. However, I'm happy to set this to zero for the gracedb pages. I don't think anything reads those values though.

File: bin/pycbc_live (hunk 7 above)
Comment (ahnitz) [reply bin/pycbc_live]: This situation doesn't occur in the current code. I.E. there are *no* followup detectors in the old sense. They are currently all used for significance if there. So the only case is if they are on or not. If they are on they are used for skymap and for significance. Currently, a single only occurs if the other detectors are offline, so there can't be a multidetector skymap.

File: bin/pycbc_live (hunk 9 above)
Comment (ahnitz) [reply bin/pycbc_live]: yes, that's right. It now knows which ifos it is supposed to process.

File: bin/pycbc_live [hunk 13]
```diff
@@ -21,6 +23,19 @@ def makedir(path):
...
 
+def ptof(p, ft):
+    return numpy.log(1.0-p) / -ft
+
+def ftop(f, ft):
+    return 1 - numpy.exp(-ft * f)
+
+def combine_ifar_pvalue(ifar, pvalue):
+    from scipy.stats import combine_pvalues
+    reference_foreground = 0.01 * ifar
```
Comment (tdent) [inline bin/pycbc_live]: need comment on what this '0.01 * ifar' does and why (if we think so) it's ok and/or desirable

**Comment by ahnitz** [comment]
@titodalcanton @tdent Remaining climate issues I don't think can be resolved.

File: bin/pycbc_live [hunk 14]
```diff
@@ -21,6 +23,19 @@ def makedir(path):
     if path is not None and not os.path.exists(path):
         os.makedirs(path)
 
+def ptof(p, ft):
+    return numpy.log(1.0-p) / -ft
```
Comment (tdent) [inline bin/pycbc_live]: PEP spacing please

File: bin/pycbc_live [hunk 15]
```diff
@@ -21,6 +23,19 @@ def makedir(path):
...
+
+def ftop(f, ft):
+    return 1 - numpy.exp(-ft * f)
+
+def combine_ifar_pvalue(ifar, pvalue):
```
Comment (tdent) [inline bin/pycbc_live]: add comment to remind us to revisit this

File: pycbc/events/coinc.py [hunk 16]
```diff
@@ -701,11 +709,62 @@ def __init__(self, num_templates, analysis_block, background_statistic,
 
         self.singles = {}
 
-        #if save_background_on_interrupt:
-        #    import signal
-        #    def sig_handler(signum, frame):
-        #        pass
-        #    signal.signal(signal.SIGINT, sig_handler)
+    # Adding this here as it requires the results format dict as defined by this
```
Comment (tdent) [inline pycbc/events/coinc.py]: remove comment - not sure it is helpful

File: pycbc/events/coinc.py [hunk 17]
```diff
@@ -701,11 +709,62 @@ def __init__(self, num_templates, analysis_block, background_statistic,
...
+        mstat = 0
+        mifar = 0
+        mresult = None
+
+        # record the trials factor from the possible coincs we could
+        # maximize over
+        trials = 0
+        for result in coinc_results:
+            # Check that a coinc was possible. This was added in the
+            # 'add_singles' method of this class
```
Comment (tdent) [inline pycbc/events/coinc.py]: 'this was added' is irrelevant - to refer people to another bit of code just say "see the add_singles method"

**Comment by ahnitz** [comment]
@tdent Ok, please take another look when you can.

**Comment by tdent** [comment]
@titodalcanton this still records you as having 'requested changes' .. is that now superseded

**Comment by titodalcanton** [comment]
I started testing this on my fake strain data. So far so good.

https://gracedb-playground.ligo.org/events/view/T8877

Nice!

**Comment by titodalcanton** [comment]
Here's one error which seems different from what @bhooshan-gadre found:
```
2019-02-20 18:45:24,674 node742 Adding V1 to candidate, pvalue 0.2520572996037793, 3280 samples
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/__main__.py", line 7, in <module>
    main()
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/run.py", line 196, in main
    run_command_line(args)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/run.py", line 47, in run_command_line
    run_path(sys.argv[0], run_name='__main__')
  File "/usr/lib64/python2.7/runpy.py", line 240, in run_path
    return _run_module_code(code, init_globals, run_name, path_name)
  File "/usr/lib64/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tito/projects/pycbc/live/multidetector/env/bin/pycbc_live", line 582, in <module>
    data_reader, bank)
  File "/home/tito/projects/pycbc/live/multidetector/env/bin/pycbc_live", line 163, in check_coincs
    bank, followup_ifos)
  File "/home/tito/projects/pycbc/live/multidetector/env/bin/pycbc_live", line 130, in compute_followup_data
    snr_series, ptime, pvalue)
  File "/home/tito/projects/pycbc/live/multidetector/env/bin/pycbc_live", line 145, in combine_far_with_followup
    triggers['foreground/{}/snr'.format(ifo)] = abs(snr_series.at_time(ptime))
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/types/timeseries.py", line 204, in at_time
    return self[int((time-self.start_time)*self.sample_rate)]
  File "</home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/decorator.pyc:decorator-gen-120>", line 2, in __getitem__
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/types/array.py", line 67, in _convert
    return fn(self, *args)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/types/array.py", line 801, in __getitem__
    return self._getvalue(index)
  File "</home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/decorator.pyc:decorator-gen-117>", line 2, in _getvalue
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/scheme.py", line 173, in scheming_function
    return _import_cache[mgr.state][fn](*args, **kwds)
  File "pycbc/types/array_cpu.pyx", line 178, in pycbc.types.array_cpu._getvalue
IndexError: index 390 is out of bounds for axis 0 with size 195
```

**Comment by tdent** [comment]
Only 5 per year FAR for the test event .. ?

**Comment by titodalcanton** [comment]
Another error came overnight as I left the analysis processing online data:
```
2019-02-20 22:22:52,292 node742 done clustering coinc triggers: 1 triggers remaining
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/__main__.py", line 7, in <module>
    main()
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/run.py", line 196, in main
    run_command_line(args)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/run.py", line 47, in run_command_line
    run_path(sys.argv[0], run_name='__main__')
  File "/usr/lib64/python2.7/runpy.py", line 240, in run_path
    return _run_module_code(code, init_globals, run_name, path_name)
  File "/usr/lib64/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tito/projects/pycbc/live/multidetector/env/bin/pycbc_live", line 573, in <module>
    coinc_results.append(c.add_singles(results, data_reader))
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/events/coinc.py", line 1179, in add_singles
    ifos=valid_ifos)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/events/coinc.py", line 1094, in _find_coincs
    self.coincs.add(cstat[cidx][bkg_idx], single_expire, ifos)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/events/coinc.py", line 720, in add
    self.timer[ifo][self.index:self.index+len(values)] = times[ifo]
ValueError: cannot copy sequence with size 4 to array axis with dimension 1
```

**Comment by titodalcanton** [comment]
My offline test completed with no errors. I am leaving the online analysis running over night to check duty factor etc but this is starting to seem mergeable!

**Comment by titodalcanton** [comment]
Nevermind, got this error again:
```
2019-02-21 21:13:16,834 node742 done clustering coinc triggers: 1 triggers remaining
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/__main__.py", line 7, in <module>
    main()
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/run.py", line 196, in main
    run_command_line(args)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/mpi4py/run.py", line 47, in run_command_line
    run_path(sys.argv[0], run_name='__main__')
  File "/usr/lib64/python2.7/runpy.py", line 240, in run_path
    return _run_module_code(code, init_globals, run_name, path_name)
  File "/usr/lib64/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/tito/projects/pycbc/live/multidetector/env/bin/pycbc_live", line 573, in <module>
    coinc_results.append(c.add_singles(results, data_reader))
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/events/coinc.py", line 1179, in add_singles
    ifos=valid_ifos)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/events/coinc.py", line 1094, in _find_coincs
    self.coincs.add(cstat[cidx][bkg_idx], single_expire, ifos)
  File "/home/tito/projects/pycbc/live/multidetector/env/lib/python2.7/site-packages/pycbc/events/coinc.py", line 720, in add
    self.timer[ifo][self.index:self.index+len(values)] = times[ifo]
ValueError: cannot copy sequence with size 11 to array axis with dimension 1
```

**Comment by titodalcanton** [comment]
The analysis I am running on commissioning data has so far accumulated tens of thousands of triggers in all combinations of 2 detectors. I think this patch can be merged and we can continue debugging further issues in separate requests.

**Comment by titodalcanton** [comment]
@ahnitz could you just rebase on master so the glue issue will not trigger Travis?

**Comment by ahnitz** [comment]
@titodalcanton Ok,

**Pull request merged**